File: oop/tests/strategies/TestMutation.java
package strategies;

import utils.RandomUtil;
import utils.FakeRandomUtil;

/**
 * javac -d bin src/models/*.java src/utils/*.java src/strategies/*.java tests/utils/FakeRandomUtil.java tests/strategies/TestMutation.java
 * java -ea -cp bin strategies.TestMutation
 */
public class TestMutation {
    /**
     * Test 01: BitFlipMutation - no mutation happens (mutation rate 0.0)
     * Tests: When mutation rate is 0.0, chromosome unchanged regardless of random values
     */
    public static void test_01() {
        boolean[] chromosome = new boolean[]{true, false, true, false, true, false, true, false, true, false};
        
        BitFlipMutation mutation = new BitFlipMutation(0.0); // Rate 0.0 = no mutations
        RandomUtil random = new RandomUtil(42); // Use real RandomUtil
        
        boolean[] mutated = mutation.mutate(chromosome, random);
        
        assert mutated.length == 10 : "Mutated chromosome should have same length";
        
        // Verify no bits were flipped
        for (int i = 0; i < 10; i++) {
            assert mutated[i] == chromosome[i] : "Bit " + i + " should remain unchanged";
        }
    }
    
    /**
     * Test 02: BitFlipMutation - all bits mutate (mutation rate 1.0)
     * Tests: When mutation rate is 1.0, all bits are always flipped
     */
    public static void test_02() {
        boolean[] chromosome = new boolean[]{true, false, true, false, true, false, true, false, true, false};
        
        BitFlipMutation mutation = new BitFlipMutation(1.0); // Rate 1.0 = all mutations
        RandomUtil random = new RandomUtil(42); // Use real RandomUtil
        
        boolean[] mutated = mutation.mutate(chromosome, random);
        
        assert mutated.length == 10 : "Mutated chromosome should have same length";
        
        // Verify all bits were flipped
        for (int i = 0; i < 10; i++) {
            assert mutated[i] == !chromosome[i] : "Bit " + i + " should be flipped";
        }
    }
    
    /**
     * Test 03: BitFlipMutation - selective mutations (determined bernoulli sequence)
     * Tests: Only specific bits mutate based on bernoulli sequence
     */
    public static void test_03() {
        boolean[] chromosome = new boolean[]{true, true, true, true, true, false, false, false, false, false};
        
        BitFlipMutation mutation = new BitFlipMutation(0.5);
        
        // FakeRandomUtil: bernoulli sequence with selective mutations
        // Mutate bits at indices: 1, 3, 5, 7, 9
        boolean[] bernoulliSeq = new boolean[]{false, true, false, true, false, true, false, true, false, true};
        FakeRandomUtil fakeRandom = new FakeRandomUtil(new int[]{}, bernoulliSeq);
        
        boolean[] mutated = mutation.mutate(chromosome, fakeRandom);
        
        assert mutated.length == 10 : "Mutated chromosome should have same length";
        
        // Verify correct bits were flipped
        assert mutated[0] == true : "Bit 0 should not mutate";
        assert mutated[1] == false : "Bit 1 should be flipped";
        assert mutated[2] == true : "Bit 2 should not mutate";
        assert mutated[3] == false : "Bit 3 should be flipped";
        assert mutated[4] == true : "Bit 4 should not mutate";
        assert mutated[5] == true : "Bit 5 should be flipped";
        assert mutated[6] == false : "Bit 6 should not mutate";
        assert mutated[7] == true : "Bit 7 should be flipped";
        assert mutated[8] == false : "Bit 8 should not mutate";
        assert mutated[9] == true : "Bit 9 should be flipped";
    }
    
    /**
     * Test 04: BitFlipMutation - edge cases
     * Tests: Null input, empty chromosome, single-bit chromosome, large chromosome, mutation rate boundaries
     */
    public static void test_04() {
        BitFlipMutation mutation = new BitFlipMutation(0.5);
        
        // Edge case 1: Null chromosome
        boolean nullCaught = false;
        try {
            mutation.mutate(null, new FakeRandomUtil(new int[]{}, new boolean[]{false}));
        } catch (IllegalArgumentException e) {
            nullCaught = true;
            assert e.getMessage().contains("null") : "Exception should mention null";
        }
        assert nullCaught : "Should throw exception for null chromosome";
        
        // Edge case 2: Empty chromosome
        boolean[] emptyChromosome = new boolean[0];
        FakeRandomUtil fakeRandom2 = new FakeRandomUtil(new int[]{}, new boolean[]{});
        boolean[] mutatedEmpty = mutation.mutate(emptyChromosome, fakeRandom2);
        assert mutatedEmpty.length == 0 : "Empty chromosome should remain empty";
        
        // Edge case 3: Single-bit chromosome
        boolean[] singleBit = new boolean[]{true};
        boolean[] bernoulliSingle = new boolean[]{true};
        FakeRandomUtil fakeRandom3 = new FakeRandomUtil(new int[]{}, bernoulliSingle);
        boolean[] mutatedSingle = mutation.mutate(singleBit, fakeRandom3);
        assert mutatedSingle.length == 1 : "Single-bit chromosome should have length 1";
        assert mutatedSingle[0] == false : "Single bit should be flipped";
        
        // Edge case 4: Large chromosome (100 bits)
        boolean[] largeChromosome = new boolean[100];
        boolean[] largeBernoulli = new boolean[100];
        for (int i = 0; i < 100; i++) {
            largeChromosome[i] = (i % 2 == 0);  // alternating pattern
            largeBernoulli[i] = (i % 3 == 0);   // mutate every 3rd bit
        }
        FakeRandomUtil fakeRandom4 = new FakeRandomUtil(new int[]{}, largeBernoulli);
        boolean[] mutatedLarge = mutation.mutate(largeChromosome, fakeRandom4);
        assert mutatedLarge.length == 100 : "Large chromosome should have correct length";
        
        // Verify correct bits were flipped in large chromosome
        for (int i = 0; i < 100; i++) {
            if (i % 3 == 0) {
                assert mutatedLarge[i] == !largeChromosome[i] : "Bit " + i + " should be flipped";
            } else {
                assert mutatedLarge[i] == largeChromosome[i] : "Bit " + i + " should not be flipped";
            }
        }
    }
    
    /**
     * Test 05: BitFlipMutation - defensive copy and independence
     * Tests: Original chromosome unchanged, mutated result is independent copy
     */
    public static void test_05() {
        boolean[] chromosome = new boolean[]{true, false, true, false, true};
        boolean[] originalCopy = new boolean[]{true, false, true, false, true};
        
        BitFlipMutation mutation = new BitFlipMutation(1.0);
        RandomUtil random = new RandomUtil(42);
        
        boolean[] mutated = mutation.mutate(chromosome, random);
        
        // Verify original chromosome unchanged
        for (int i = 0; i < 5; i++) {
            assert chromosome[i] == originalCopy[i] : "Original chromosome at " + i + " should not be modified";
        }
        
        // Verify mutated is a different array
        assert mutated != chromosome : "Mutated should be a different array object";
        
        // Modify mutated array and verify original is unaffected
        mutated[0] = true;
        assert chromosome[0] == originalCopy[0] : "Original chromosome should be independent from mutated result";
    }
    
    public static void main(String[] args) {
        test_01();
        test_02();
        test_03();
        test_04();
        test_05();
        System.out.println("All mutation tests passed!");
    }
}

File: fp/tests/test_mutation.py
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent / "src"))

from ga_operators import bitflip_mutate
from type_alias import GeneString
from fake_random_gen import fake_random_engine


class TestMutation:
    """Test suite for bitflip_mutate operator"""
    
    @staticmethod
    def test_01():
        """Test 01: BitFlipMutation - no mutation happens (mutation rate 0.0)
        Tests: When mutation rate is 0.0, chromosome unchanged regardless of random values
        """
        chromosome = [True, False, True, False, True, False, True, False, True, False]
        
        # With mutation rate 0.0, bernoulli always returns False (no flips)
        # Need 10 bernoulli False values (2**256 in fake_random_engine)
        rand_eng = fake_random_engine([2**256] * 10)
        mutated = bitflip_mutate(chromosome, 0.0, rand_eng=rand_eng)
        
        assert len(mutated) == 10, "Mutated chromosome should have same length"
        assert mutated == chromosome, "Chromosome should be unchanged with rate 0.0"
    
    @staticmethod
    def test_02():
        """Test 02: BitFlipMutation - all bits mutate (mutation rate 1.0)
        Tests: When mutation rate is 1.0, all bits are always flipped
        """
        chromosome = [True, False, True, False, True, False, True, False, True, False]
        
        # With mutation rate 1.0, every bernoulli returns True (flip)
        # Need 10 bernoulli True values (0 in fake_random_engine)
        rand_eng = fake_random_engine([0] * 10)
        mutated = bitflip_mutate(chromosome, 1.0, rand_eng=rand_eng)
        
        assert len(mutated) == 10, "Mutated chromosome should have same length"
        # All bits should be flipped
        for i in range(10):
            assert mutated[i] == (not chromosome[i]), f"Bit {i} should be flipped"
    
    @staticmethod
    def test_03():
        """Test 03: BitFlipMutation - selective mutations (determined bernoulli sequence)
        Tests: Only specific bits mutate based on bernoulli sequence
        """
        chromosome = [True, True, True, True, True, False, False, False, False, False]
        
        # Mutation rate 0.5, flip bits at indices 1, 3, 5, 7, 9
        # Bernoulli sequence: [False(no), True(yes), False(no), True(yes), ...]
        # 0 means bernoulli returns True (flip), 2**256 means False (no flip)
        bernoulli_seq = [2**256, 0, 2**256, 0, 2**256, 0, 2**256, 0, 2**256, 0]
        rand_eng = fake_random_engine(bernoulli_seq)
        mutated = bitflip_mutate(chromosome, 0.5, rand_eng=rand_eng)
        
        assert len(mutated) == 10, "Mutated chromosome should have same length"
        
        # Verify correct bits were flipped
        assert mutated[0] == True, "Bit 0 should not mutate"
        assert mutated[1] == False, "Bit 1 should be flipped"
        assert mutated[2] == True, "Bit 2 should not mutate"
        assert mutated[3] == False, "Bit 3 should be flipped"
        assert mutated[4] == True, "Bit 4 should not mutate"
        assert mutated[5] == True, "Bit 5 should be flipped"
        assert mutated[6] == False, "Bit 6 should not mutate"
        assert mutated[7] == True, "Bit 7 should be flipped"
        assert mutated[8] == False, "Bit 8 should not mutate"
        assert mutated[9] == True, "Bit 9 should be flipped"
    
    @staticmethod
    def test_04():
        """Test 04: BitFlipMutation - edge cases
        Tests: Empty chromosome, single-bit chromosome, large chromosome
        """
        # Edge case 1: Empty chromosome
        empty_chromosome: GeneString = []
        rand_eng = fake_random_engine([])
        mutated_empty = bitflip_mutate(empty_chromosome, 0.5, rand_eng=rand_eng)
        assert len(mutated_empty) == 0, "Empty chromosome should remain empty"
        
        # Edge case 2: Single-bit chromosome (all mutation)
        single_bit: GeneString = [True]
        rand_eng = fake_random_engine([0])  # Bernoulli True (flip)
        mutated_single = bitflip_mutate(single_bit, 1.0, rand_eng=rand_eng)
        assert len(mutated_single) == 1, "Single-bit chromosome should have length 1"
        assert mutated_single[0] == False, "Single bit should be flipped"
        
        # Edge case 3: Large chromosome (100 bits)
        large_chromosome: GeneString = [i % 2 == 0 for i in range(100)]  # alternating pattern
        # Mutation at every 3rd bit (indices 0, 3, 6, 9, ...)
        bernoulli_large = []
        for i in range(100):
            if i % 3 == 0:
                bernoulli_large.append(0)  # Flip
            else:
                bernoulli_large.append(2**256)  # No flip
        
        rand_eng = fake_random_engine(bernoulli_large)
        mutated_large = bitflip_mutate(large_chromosome, 0.5, rand_eng=rand_eng)
        assert len(mutated_large) == 100, "Large chromosome should have correct length"
        
        # Verify correct bits were flipped in large chromosome
        for i in range(100):
            if i % 3 == 0:
                assert mutated_large[i] == (not large_chromosome[i]), f"Bit {i} should be flipped"
            else:
                assert mutated_large[i] == large_chromosome[i], f"Bit {i} should not be flipped"
    
    @staticmethod
    def test_05():
        """Test 05: BitFlipMutation - original chromosome unchanged
        Tests: Original chromosome is not modified (defensive copy)
        """
        chromosome = [True, False, True, False, True]
        original_copy = chromosome.copy()
        
        # All bits mutate
        rand_eng = fake_random_engine([0] * 5)
        mutated = bitflip_mutate(chromosome, 1.0, rand_eng=rand_eng)
        
        # Verify original chromosome unchanged
        assert chromosome == original_copy, "Original chromosome should not be modified"
        
        # Verify mutated is different
        assert mutated != chromosome, "Mutated should be different from original"
        
        # Verify mutated is independent
        mutated[0] = True
        assert chromosome[0] == original_copy[0], "Original chromosome should be independent from mutated result"
    
    @staticmethod
    def test_06():
        """Test 06: BitFlipMutation - two-bit chromosome with mixed mutations
        Tests: Verify mutation works correctly with minimal chromosome size
        """
        chromosome = [True, True]
        
        # First bit mutates, second bit doesn't
        rand_eng = fake_random_engine([0, 2**256])
        mutated = bitflip_mutate(chromosome, 0.5, rand_eng=rand_eng)
        
        assert len(mutated) == 2, "Should have length 2"
        assert mutated[0] == False, "First bit should be flipped"
        assert mutated[1] == True, "Second bit should not be flipped"
        assert chromosome == [True, True], "FP: Original chromosome not modified by mutation"


def run_all_tests():
    """Run all test methods"""
    print("Running Mutation tests...")
    TestMutation.test_01()
    print("  ✓ test_01: No mutation (rate 0.0)")
    TestMutation.test_02()
    print("  ✓ test_02: All bits mutate (rate 1.0)")
    TestMutation.test_03()
    print("  ✓ test_03: Selective mutations")
    TestMutation.test_04()
    print("  ✓ test_04: Edge cases (empty, single-bit, large)")
    TestMutation.test_05()
    print("  ✓ test_05: Original chromosome unchanged")
    TestMutation.test_06()
    print("  ✓ test_06: Two-bit chromosome")
    
    print("\n✅ All mutation tests passed!")


if __name__ == "__main__":
    run_all_tests()
